File: backend/app/routes/audit_logs.py
from datetime import datetime, timezone, timedelta
from fastapi import APIRouter, Depends, Query

from app.core.database import audit_logs_collection
from app.services.dependencies import get_current_user

router = APIRouter(prefix="/api/audit-logs", tags=["audit-logs"])


async def log_action(user: dict, action: str, details: str) -> None:
    """
    Mirrors logAction from auditLogs.js. Fire-and-forget by design —
    original wraps this in try/except and only console.errors on failure,
    never lets a logging failure break the actual write it's attached to.
    Callers should NOT await-and-raise; call this, don't let it block the
    real response if it fails.
    """
    try:
        user_email = (user or {}).get("email") or "Unknown/System"
        await audit_logs_collection.insert_one({
            "user": user_email,
            "action": action,
            "details": details,
            "timestamp": datetime.now(timezone.utc),
        })
    except Exception as e:
        print(f"Error logging action: {e}")


@router.get("")
async def get_recent_logs(days: int = Query(30), user: dict = Depends(get_current_user)):
    past_date = datetime.now(timezone.utc) - timedelta(days=days)
    cursor = audit_logs_collection.find(
        {"timestamp": {"$gte": past_date}}
    ).sort("timestamp", -1)

    docs = await cursor.to_list(length=None)
    for d in docs:
        d["id"] = str(d.pop("_id"))
    return docs

File: client/src/firebase/services/auditLogs.js
import { apiFetch } from "../api";

/**
 * No-op: the backend logs actions server-side on every write operation.
 * Kept as export so existing callers don't break.
 */
export const logAction = async (action, details) => {
  // Server handles audit logging automatically
};

export const getRecentLogs = async (days = 30) => {
  try {
    const logs = await apiFetch(`/api/audit-logs?days=${days}`);
    return (logs || []).map(log => ({
      ...log,
      timestamp: log.timestamp ? new Date(log.timestamp) : new Date(),
    }));
  } catch (error) {
    console.error("Error fetching audit logs:", error);
    return [];
  }
};
